source/module_lr/potentials/xc_kernel.cpp

```diff
@@ -0,0 +1,221 @@
+    }
+
+    // -----------------------------------------------------------------------------------
+    // for GGA
+    const bool is_gga = [&funcs]()
+        {
+            for (xc_func_type& func : funcs)
+            {
+                switch (func.info->family)
+                {
+                case XC_FAMILY_GGA:
+                case XC_FAMILY_HYB_GGA:
+                    return true;
+                }
+            }
+            return false;
+        }();
+    std::vector<std::vector<ModuleBase::Vector3<double>>> gdr;  // \nabla \rho
+    std::vector<double> sigma;  // |\nabla\rho|^2
+    std::vector<double> sgn;        // sgn for threshold mask
+    if (is_gga)
+    {
+        // 0. set up sgn for threshold mask
+        // in the case of GGA correlation for polarized case,
+        // a cutoff for grho is required to ensure that libxc gives reasonable results
+
+        // 1. \nabla \rho
+        gdr.resize(nspin);
+        for (int is = 0; is < nspin; ++is)
+        {
+            std::vector<double> rhor(nrxx);
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static, 1024)
+#endif
+            for (int ir = 0; ir < nrxx; ++ir) rhor[ir] = rho[ir * nspin + is];
+            gdr[is].resize(nrxx);
+            LR_Util::grad(rhor.data(), gdr[is].data(), rho_basis_, tpiba);
+        }
+        // 2. |\nabla\rho|^2
+        sigma.resize(nrxx * ((1 == nspin) ? 1 : 3));
+        if (1 == nspin)
+        {
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static, 1024)
+#endif
+            for (int ir = 0; ir < nrxx; ++ir)
+                sigma[ir] = gdr[0][ir] * gdr[0][ir];
+        }
+        else
+        {
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static, 256)
+#endif
+            for (int ir = 0; ir < nrxx; ++ir)
+            {
+                sigma[ir * 3] = gdr[0][ir] * gdr[0][ir];
+                sigma[ir * 3 + 1] = gdr[0][ir] * gdr[1][ir];
+                sigma[ir * 3 + 2] = gdr[1][ir] * gdr[1][ir];
+            }
+        }
+    }
+    // -----------------------------------------------------------------------------------
+    //==================== XC Kernels (f_xc)=============================
+    this->kernel_set_.emplace("vrho", std::vector<double>(nspin * nrxx));
+    this->kernel_set_.emplace("v2rho2", std::vector<double>(((1 == nspin) ? 1 : 3) * nrxx));//(nrxx* ((1 == nspin) ? 1 : 3)): 00, 01, 11
+    if (is_gga)
+    {
+        this->kernel_set_.emplace("vsigma", std::vector<double>(((1 == nspin) ? 1 : 3) * nrxx)); //(nrxx*): 2 for rho * 3 for sigma: 00, 01, 02, 10, 11, 12
+        this->kernel_set_.emplace("v2rhosigma", std::vector<double>(((1 == nspin) ? 1 : 6) * nrxx)); //(nrxx*): 2 for rho * 3 for sigma: 00, 01, 02, 10, 11, 12
+        this->kernel_set_.emplace("v2sigma2", std::vector<double>(((1 == nspin) ? 1 : 6) * nrxx));   //(nrxx* ((1 == nspin) ? 1 : 6)): 00, 01, 02, 11, 12, 22
+    }
```

Comment:
Yes, they are necessary to co-exist. 
Using map is only because the string-type key seems clearer than the index of a vector.
There are at most 4 keys (vsigma, v2rho2, v2rhosigma and v2sigma2) now in `kernel_set_`, so the time complexity does not count. 
But in the near future there will be at most 20 keys in `grad_kernel_set_`, I will consider orgnize them as vector...

Reply:
OK. I've changed the map into several member variables in the latest commit.